core/src/test/scala/unit/kafka/log/LogCleanerTest.scala

```diff
@@ -1214,6 +1210,54 @@ class LogCleanerTest extends JUnitSuite {
     assertEquals("The tombstone should be retained.", 1, log.logSegments.head.log.batches.iterator.next().lastOffset)
   }
 
+  /**
+    * Verify that cleaner does not crash, and handles the case where a log segment has messages
+    * with offsets that overflow base_offset + Integer.MAX_VALUE due to KAFKA-5413.
+    */
+  @Test
+  def testLogSegmentWithIndexOffsetOverflow(): Unit = {
+    val tp = new TopicPartition("test", 0)
+    val cleaner = makeCleaner(Int.MaxValue)
```

Comment:
We should probably configure the log with a low `log.index.interval.bytes` so that every append forces an attempt to write to the index.